zerver/models/streams.py

```diff
@@ -400,3 +390,24 @@ def to_dict(self) -> dict[str, Any]:
 
 def get_default_stream_groups(realm: Realm) -> QuerySet[DefaultStreamGroup]:
     return DefaultStreamGroup.objects.filter(realm=realm)
+
+
+class ChannelEmailAddress(models.Model):
+    channel = models.ForeignKey(Stream, on_delete=CASCADE)
+    creator = models.ForeignKey(UserProfile, null=True, on_delete=CASCADE, related_name="+")
+    sender = models.ForeignKey(UserProfile, on_delete=CASCADE, related_name="+")
+
+    # Used by the e-mail forwarder. The e-mail RFC specifies a maximum
+    # e-mail length of 254, and our max stream length is 30, so we
+    # have plenty of room for the token.
+    email_token = models.CharField(
+        max_length=32,
+        default=generate_email_token_for_stream,
+        unique=True,
+    )
+
+    date_created = models.DateTimeField(default=timezone_now)
+    deactivated = models.BooleanField(default=False)
```

Comment:
`MissedMessageEmailAddress` is a fairly similar table; looking at it and reflecting a bit, I think:
-  `email_token` should have `db_index=True`. I think `unique=True` makes an index anyway, but it's probably good to be explicit.
- `realm` would be a good column to have -- we've generally been including that in most tables for sharding convenience.

Reply:
Hmm, the main lookup in `decode_stream_email_address` doesn't have access to the realm, so realistically we're doing lookups just by `token` for now.

Reply:
* Added `realm` column
* I ended up adding an index on `realm, channel`. It will help in the newly added `get_or_create` query in `encode_email_address` and `.filter(...).update(deactivated=True/False)` query in `do_deactivate_stream` / `do_unarchive_stream`.

Tested this query after creating 521 streams :
```py
    channel_email_address, ignored = ChannelEmailAddress.objects.get_or_create(
        realm=stream.realm,
        channel=stream,
        creator=stream.creator,
        sender=get_system_bot(settings.EMAIL_GATEWAY_BOT, stream.realm_id),
    )
```

```zsh

zulip=> explain analyze SELECT id, realm_id, channel_id, creator_id, sender_id, email_token, date_created, deactivated FROM zerver_channelemailaddress WHERE (channel_id = 4 AND creator_id IS NULL AND realm_id = 2 AND sender_id = 1);
                                                                                   QUERY PLAN              
                                                                     
-----------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------
 Index Scan using zerver_channelemailaddress_realm_id_channel_id_idx on zerver_channelemailaddress  (cost=0
.28..8.30 rows=1 width=70) (actual time=0.171..0.172 rows=1 loops=1)
   Index Cond: ((realm_id = 2) AND (channel_id = 4))
   Filter: ((creator_id IS NULL) AND (sender_id = 1))
 Planning Time: 0.907 ms
 Execution Time: 0.225 ms
(5 rows)
```

Earlier, it was a sequential scan (as I made an incorrect assumption of channels count being not huge)

Reply:
Sounds good. Generally we should never skip adding an index when it can avoid sequential scans for production code( migrations can be different); they'll always be bad at some scale, and it's best to not leave land mines for ourselves.